### PR TITLE
Introduce merge join in query plan

### PR DESCRIPTION
**Pull request by kewang1024**
```
== RELEASE NOTE ==
Hive Changes
* Replace hive session property from  ``streaming_aggregation_enabled`` to  ``order_based_execution_enabled``
* Replace hive config from ``hive.streaming-aggregation-enabled`` to ``hive.order-based-execution-enabled``

```

Note that this change only annotates the JoinNode to indicate merge join is enabled, the execution side change on merge join is not implemented yet. 
At the moment even when merge join is enabled in the plan, it has no execution side impact.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java:274]: We may need to only do this when the merge join is enabled.
Maybe add a comment:
This is to avoid adding an exchange between the join and scan when the merge join is enabled.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:481]: Maybe change this to PreferMergeJoin as it cannot be enabled in execution yet in the java step.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java
Comment (yuanzhanhku) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1367]: And the ordering property of the table is exposed to the optimizer to plan streaming aggregation and merge join when possible.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java
Comment (yuanzhanhku) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java:421]: And the ordering property of the table is exposed to the optimizer to plan streaming aggregation and merge join when possible.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: order_based_execution_enabled must be enabled to make merge join planning work.

**Review by yuanzhanhku — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: I'm not sure if we should mention that given `order_based_execution_enabled ` would be a connector session property. And from engine level, it shouldn't indicate details on how connector implements its feature.

Thus I put the abstract and common requirement here: sorted join inputs

**Comment by kewang1024** [comment]
Hey, @yuanzhanhku 
For the logic on how we should avoid adding LocalExchange on top of right side TableScan 
the original implementation was a bit hacky so I changed it after your final review, let me know if you have concerns with the newest change

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:737]: It seems like this change should belong in a new optimizer rule (similar to `IndexJoinOptimizer`)?  I.e. I'm not sure how this relates to adding local exchange?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:157]: Comment goes above

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:752]: the new node would be further passed to the returning statement `rebaseAndDeriveProperties(node, ImmutableList.of(probe, build))`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:737]: So basically we want to eliminate adding local exchange on top of TableScan for merge join

If we have hash based join, the plan would be
JoinNode
- TableScanA
- LocalExchange
   - TableScanB

But if the table property layout can meet merge join requirements, we don't need to add the LocalExchange, it would be
JoinNode (merge join)
- TableScanA
- TableScanB

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:737]: My point is the introduction of JoinNode (merge join) seems to be more appropriately done in a separate optimizer rule, with this one simply removing the local exchange as you say.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:752]: Got it, I see that now.  As I mentioned before, should we move this to its own rule?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java]: Just move it to a new Optimizer rule, can you help review?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: I think we need to document somewhere about what are all the session properties required to enable this feature. Otherwise, other developers/users will have a hard time to figure it out. This seems to be a good place to add this documentation as this session property name sounds like it will enable merge join. If you have a better place to document this, that is also fine.

> And from engine level, it shouldn't indicate details on how connector implements its feature.

For code/design perspective, I definitely agree to have a good abstraction. But for comments, I think it is better to document all the assumptions. E.g., we could add note here to say that currently this is only supported for pre-sorted hive tables and requires enabling order_based_execution_enabled to make it work.

**Comment by yuanzhanhku** [comment]
> Hey, @yuanzhanhku For the logic on how we should avoid adding LocalExchange on top of right side TableScan the original implementation was a bit hacky so I changed it after your final review, let me know if you have concerns with the newest change

Looks good. Moving it to a new rule is cleaner. Thanks Tim for the suggestion and Thanks Ke for making this change.

**Review by yuanzhanhku — APPROVED** [review]

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java
Comment (tdcmeehan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1366]: Anyone reasonably expected to be using this config?  In which case, we'll want to deprecate the old naming.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LocalProperties.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LocalProperties.java:54]:
```suggestion
        return columns.stream().map(column -> new SortingProperty<>(column, order)).collect(toImmutableList());
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:90]: Should we consider adding a new JoinNode type?  Some aspects of `JoinNode` seem specific to hash join, like `dynamicFilters`.  `DistributionType` also seems like it might not make much sense, since we wouldn't have broadcast join for merge join (right?).  For reference, there is a separate `IndexJoinNode` type for index join.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:629]: The comment above seems to indicate we shouldn't change the plan shape.  But I believe we are, since we're eliminating exchanges?  Should we consider moving this before `AddExchanges`, then performing a minor alteration on `AddExchanges` to skip over merge join nodes?

**Review by tdcmeehan — COMMENTED** [review]
Just a few more comments, the rest looks good.

**Review by highker — COMMENTED** [review]
OK, I will remove myself from review. Thanks @tdcmeehan for helping reviewing/merging the PR :D

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java
Comment (yuanzhanhku) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1366]: This will be shared by both the streaming aggregation and merge join feature to expose the ordering property of the table and disable file splitting.
The original naming is confusing if we also use it for merge join.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:90]: This is a good point. Eventually, I think we do want to add a new JoinNode for merge join. But currently it is not supported on the execution side yet. Reusing the existing JoinNode allows the worker to fall back to use Hash Join operator when merge join operator is not implemented. 

I think maybe we can work on introducing a new JoinNode and remove the fall back path when both Java and CPP workers add the merge join implementations. What do you think, Tim?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:629]: Good point. Does it work if we add this rule right after the AddLocalExchanges rule? In that case, we don't need to update the AddLocalExchanges rule.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: Why do we need an extra property order_based_execution_enabled?  Can we always use order based execution if it's supported by the table and prefer_merge_join is true.

Also, if there's no execution support yet for this property we should set it to hidden.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java:82]: I don't think mergeJoinEnabled is the right name here.  That sounds like a session property, where as this should be actually setting a different join type.    isMergeJoin would be a better name.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:629]: I strongly agree with @tdcmeehan.  I think the structure should be as follows

1. Add rule to turn hash join -> merge join
2. Change AddLocalExchanges not to add a localExchange for merge join.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:90]: I would say we shouldn't be planning a merge join if it's not supported in the execution, and we should fail in the execution if we create a mergejoin in the planner and it's not supported in the execution yet.   Otherwise I feel we are just asking for problems.  e.g. here we're removing a local exchange when we have mergeJoinEnabled,even though we would want that local exchange if we were to fall back to a hash join.

Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:90]: I think we agree on the final state to add a new JoinNode. I don't have a strong preference on the intermediate state about whether to reject the query or fall back to hash join, as we are not supposed to enable it before the execution part is ready.
Rejecting the query seems a bit cleaner. We can go with that approach.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:629]: I agree. This structure is cleaner.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: I don't think setting it to "hidden" will prevent you from setting it in a cpp cluster.  It's more about revealing it to users.  It means it doesn't show up if you say "show session".  

I think it's totally fine to have one property for streaming aggregation and one for streaming join.  I'm just thinking in general that features are easier to use if you only have to set one property to turn them on.  Otherwise, it's easy to have mistakes where someone sets one and thinks they've enabled the feature because they don't realize you need to set both. Having two properties to enable two different features is easier to reason about as long as the names make it clear what they do.(e.g. `prefer_streaming_aggregation` and `prefer_streaming_join`). 

That being said if you would expect users to want to turn on/off streaming stuff all in one go, then that could be a reason to introduce the master property.  e.g. for spill we have the `spill_enabled` property, since in general you either want spill on or off, but we also have properties for disabling spill for individual operators in case there are bugs or issues with specific operator implementations. So for this case you could have one property `prefer_streaming_operators` and then other properties, `streaming_aggregation_enabled` and `streaming_join_enabled`.  I think a key there is having the operator specific knobs default to true once there are tested java implementations so that users mostly still just have to think about one property.  I still think just having separate properties is simpler though

What do you think?

Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: Setting the property to be hidden
Adding the description on order_based_execution_enabled as well

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:629]: I remember we used to say the plan is completely logical before AddExchange and AddLocalExchange, and we don't access data layout properties until those two optimization rule. 
After that, the plan becomes physical.

However MergeJoinOptimizer has to access all the data layout and if we move it before AddLocalExchanges, it changes all the assumptions above

another example is `RuntimeReorderJoinSides`, it's also placed after `AddLocalExchange`

thoughts @tdcmeehan @rschlussel

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: @rschlussel 
The reason why we can't always enable order_based_execution_enabled is that it would cause performance regression without working together with merge join or streaming aggregation

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:90]: How about the implementation on CPP side as of now? I remember they were using the original JoinNode

To unblock the current test, as of now, we can add it in java execution side when we see mergeJoinEnabled true, we reject the query. 

Then when we move to a new JoinNode type as a final state, we make changes to plan side and execution side from both CPP and Java

@yuanzhanhku @rschlussel thoughts?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1171]: I apologize, I think I misunderstood the situation before.  Am I understanding correctly as follows: `order_based_execution_enabled` is a property of the hive connector that determines whether to reveal the underlying sortedness of the table to the optimizer via the TableLayout. `prefer_merge_join` is a flag to enable/disable the merge join optimizer.  That seems totally fine to me.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:90]: CPP side needs changes anyway for either approach. We have a cpp side prototype. But that is not submitted yet. This planning side change needs to go in first such that the CPP side can create the merge join operator according to the plan change. I think the cpp side complexity of the two approaches are similar, i.e., create the merge join operator if the new bool mergeJoinEnabled is set to true in the original join node or if it is a new join node.

So the cpp side change is not a blocker for introducing a new join node type.

**Comment by kewang1024** [comment]
I have some difficulty changing the optimizer to a Rule, currently still working on it @rschlussel 


Can we first review the rest to see if they look good?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java
Comment (rschlussel) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java:171]: can you add this test?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java
Comment (rschlussel) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1366]: Can you also add documentation for this to the Hive Connector docs?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1179]: Originally it's written in that way and I mentioned the same logic to @yuanzhanhku, but he had strong opinion on adding the hive connector note, see the discussion above (I just un-resolve the comment above)

@yuanzhanhku let's make a final decision here on how the note should be?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java
Comment (rschlussel) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java:171]: got it. Can you add a test for whatever the expected result is for multiple join keys, and then we can update once support is added?  Just want to make sure all behaviors are well defined.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:596]: Essentially MergeJoinOptimizer can only eliminate the local exchange in a JoinNode and convert it to MergeJoinNode

We still rely on AddExchange to first divide the plan to different stages, for example 
JoinNode
-TableScan A ( bucketed by a, sorted by c)
-TableScan B ( bucketed by b, sorted by c)

MergeJoinOptimizer only compares the local sorting properties and see if they're sorted by the join key. Thus it would think it could optimize the plan which is actually wrong because A and B are bucketed by different keys 
We need to rely on AddExchanges to change the plan to 
JoinNode
-Exchange
      -TableScan A ( bucketed by a, sorted by c)
-TableScan B ( bucketed by b, sorted by c)


moving MergeJoinOptimizer will introduce a lot of unnecessary work,

File: presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java
Comment (kewang1024) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestMergeJoinPlan.java:94]: both
1. because it's not bucketed on the same key as the other table, two scans can't be in the same stage, there would be Exchange
2. If they're not bucketed on join key, two scans can't be in the same stage either

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:98]: So the requirements are not only sorted by join keys,  we plan to support more scenarios where join keys are only prefix of order keys so we put a more abstract name here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (kewang1024) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java]: If A is ordered by a, b; and B is ordered by b, a
We actually can't enable merge join for them

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1179]: Since it requires two session properties to make it work for hive connector. I think it is better to document this somewhere. Otherwise, people will not be able to figure it out. Maybe have a post when this feature is ready or document it in some code comment is also fine.  If we can find a better place to document this, we can remove it here.

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1179]: Maybe a good compromise would be a generic note about connectors exposing sortedness here, and when the feature is implemented we can add formal documentation for how to use it.

File: presto-docs/src/main/sphinx/optimizer/merge-join-optimizations.rst
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/merge-join-optimizations.rst:1]: The documentation is great, but I think we should wait to add it until the feature is ready to use to before adding it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:596]: Thanks for the explanation!  Makes a lot of sense. Maybe later if we want to expand usage, we can think about how to move it up.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:97]: That's fine. If you want a shorter name you can do something like `isMergeJoinEligible`. 
 If not, just fix the typo.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1179]: Sounds good to me.

**Comment by kewang1024** [comment]
Thanks guys for the review, can you help take a final look? @rschlussel @yuanzhanhku

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java
Comment (yuanzhanhku) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java:95]: Merge join doesn't require both sides to be table scans.
For example, we can have
MergeJoinNode
- StreamingAggregation
  - TableScan
- Sort
  - TableScan 

There could be projects/StreamingAggregation/Sort or another merge join above the TableScan.
As long as the child subplan can guarantee the sorted property, we should be able to enable merge join.

**Comment by rschlussel** [comment]
Looks like there is a checkstyle violation, but otherwise looks good!

**Review by yuanzhanhku — APPROVED** [review]

**Review by rschlussel — APPROVED** [review]

**Comment by sleep1661** [comment]
@kewang1024 Merge-Join is a valuable feature. And I want to know if the execution of merge join was implemented now,  or any plan for it?

**Pull request merged**
